test_runner/regress/test_branching.py

```diff
@@ -128,3 +137,233 @@ def test_branching_unnormalized_start_lsn(neon_simple_env: NeonEnv, pg_bin: PgBi
+
+
+def test_cannot_branch_from_non_uploaded_branch(neon_env_builder: NeonEnvBuilder):
+    """
+    Branch should not be possible to create because ancestor has not been uploaded.
+    """
+
+    env = neon_env_builder.init_configs()
+    env.start()
+
+    env.pageserver.allowed_errors.append(
+        ".*request{method=POST path=/v1/tenant/.*/timeline request_id=.*}: request was dropped before completing.*"
+    )
+    ps_http = env.pageserver.http_client()
+
+    # pause all uploads
+    ps_http.configure_failpoints(("before-upload-index-pausable", "pause"))
+    ps_http.tenant_create(env.initial_tenant)
+
+    with pytest.raises(ReadTimeout):
+        ps_http.timeline_create(env.pg_version, env.initial_tenant, env.initial_timeline, timeout=2)
+
+    branch_id = TimelineId.generate()
+
+    with pytest.raises(PageserverApiException, match="ancestor timeline is not active"):
+        ps_http.timeline_create(
+            env.pg_version,
+            env.initial_tenant,
+            branch_id,
+            ancestor_timeline_id=env.initial_timeline,
+        )
+
+    with pytest.raises(
+        PageserverApiException,
+        match=f"NotFound: Timeline {env.initial_tenant}/{branch_id} was not found",
+    ):
+        ps_http.timeline_detail(env.initial_tenant, branch_id)
+        # important to note that a task might still be in progress to complete
+        # the work, but will never get to that because we have the pause
+        # failpoint
+
+    # FIXME: paused uploads bother shutdown
+    env.pageserver.stop(immediate=True)
+
+
+def test_competing_branchings_from_loading_race_to_ok_or_err(neon_env_builder: NeonEnvBuilder):
+    """
+    If the activate only after upload is used, then retries could become competing.
+    """
+
+    env = neon_env_builder.init_configs()
+    env.start()
+
+    env.pageserver.allowed_errors.append(
+        ".*request{method=POST path=/v1/tenant/.*/timeline request_id=.*}: request was dropped before completing.*"
+    )
+    env.pageserver.allowed_errors.append(
+        ".*Error processing HTTP request: InternalServerError\\(Timeline .*/.* already exists in pageserver's memory"
+    )
+    ps_http = env.pageserver.http_client()
+
+    # pause all uploads
+    ps_http.configure_failpoints(("before-upload-index-pausable", "pause"))
+    ps_http.tenant_create(env.initial_tenant)
+
+    with pytest.raises(ReadTimeout):
+        ps_http.timeline_create(env.pg_version, env.initial_tenant, env.initial_timeline, timeout=2)
+
+    branch_id = TimelineId.generate()
+
+    queue: SimpleQueue[Union[Dict[Any, Any], Exception]] = SimpleQueue()
+    barrier = threading.Barrier(3)
+
+    def try_branch():
+        barrier.wait()
+        barrier.wait()
+        try:
+            ret = ps_http.timeline_create(
+                env.pg_version,
+                env.initial_tenant,
+                branch_id,
+                ancestor_timeline_id=env.initial_timeline,
+                timeout=5,
```

Comment:
Perhaps this should not have a timeout either, given the recent flakyness.

Reply:
[cb2e4f4](https://github.com/neondatabase/neon/pull/5484/commits/cb2e4f4f8d25fc06f4fb441410dc93c425f13240)